src/driver/Firestore/InProcessFirestore.ts

```diff
@@ -1340,4 +1347,60 @@ class InProcessFirestoreBulkWriter implements IFirestoreBulkWriter {
             )
         }
     }
+    
+    private enqueue(op: () => Promise<any>): Promise<any> {
+        const bulkWriterOp = new BulkWriterOperation(op)
+        this.writeOperations.push(bulkWriterOp)
+        return bulkWriterOp.promise
+    }
 }
+
+/**
+ * Represents a single write for BulkWriter, encapsulating operation execution
+ * and error handling.
+ */
+ class BulkWriterOperation {
+    deferred: Deferred
+    op: () => Promise<IFirestoreWriteResult>
+    
+    constructor(op: () => Promise<IFirestoreWriteResult>) {
+        this.op = op
+        this.deferred = new Deferred();
+    }
+
+    get promise() {
+        return this.deferred.promise;
+    }
+
+    onError(error: Error) {
+        this.deferred.reject(error)
+    }
+
+    onSuccess(result: IFirestoreWriteResult){
+        this.deferred.resolve(result)
+    }
+}
+
+/**
+ * A Promise implementation that supports deferred resolution.
+ * Source: BulkWriter implementation
+ */
+ class Deferred {
+    resolve: (value: unknown) => void
+    reject: (reason: unknown) => void
+    promise: Promise<unknown>
+
+    constructor() {
+        this.resolve = () => { }
+        this.reject = () => { }
+        this.promise = new Promise((resolve, reject) => {
+            this.resolve = resolve
+            this.reject = reject
+        });
+    }
+}
+
+// Swallows errors
```

Comment:
Some more info about why we need this would be useful for future archaeologists 🕵️‍♀️

Reply:
added more references to the code/docs hopefully it's clearer